feat(deployments): add data source

mitchnielsen opened the pull request:
Implements the data source for Deployments.

Closes https://linear.app/prefect/issue/PLA-799/add-datasource-support-for-deployments

Closes https://github.com/PrefectHQ/terraform-provider-prefect/issues/314

mitchnielsen committed:
feat(deployments): add data source

Implements the data source for Deployments.

Closes https://linear.app/prefect/issue/PLA-799/add-datasource-support-for-deployments

Closes https://github.com/PrefectHQ/terraform-provider-prefect/issues/314

mitchnielsen committed:
Add the datasource to the provider

mitchnielsen committed:
Add tests for deployment datasource

mitchnielsen commented on internal/provider/datasources/deployment.go:
```diff
@@ -0,0 +1,405 @@
+- deployment name
+The Deployment ID takes precedence over deployment name.
+`,
+		Attributes: map[string]schema.Attribute{
```
I'd love to just copy the attributes from the Deployment resource, but the fields on each attribute are a little different for data sources.

mitchnielsen commented on internal/provider/datasources/deployment.go:
```diff
@@ -0,0 +1,405 @@
+	client api.PrefectClient
+}
+
+// DeploymentDataSourceModel defines the Terraform data source model.
+type DeploymentDataSourceModel struct {
+	// The model requires the same fields, so reuse the fields defined in the resource model.
+	resources.DeploymentResourceModel
+}
```
afaik, this is the first time trying something like this - let me know if there are any drawbacks I may have missed

parkedwards replied:
honestly, im not sure there is a drawback yet. they use the same `types` package

mitchnielsen replied:
was thinking we could make a new `common` package next to `resources` and `datasources` and put common elements there 🤔

mitchnielsen commented on internal/provider/datasources/deployment.go:
```diff
@@ -0,0 +1,405 @@
+	d.client = client
+}
+
+// copyDeploymentToModel leverages the function by the same name from the resources package to avoid repeating
+// the logic. Because DeploymentResourceModel embeds the resource.DeploymentResourceModel type, we can cast
+// it to the compatible type before calling the referenced function.
+func copyDeploymentToModel(ctx context.Context, deployment *api.Deployment, model *DeploymentDataSourceModel) diag.Diagnostics {
```
This accompanies the embedded struct for the data source model. It's a bit clunky to translate between the two types, but imo it beats repeating all of the logic in CopyDeploymentToModel. open to feedback though

parkedwards replied:
well this is pretty smart

mitchnielsen committed:
Add example, docs for deployment datasource

parkedwards commented on internal/provider/datasources/deployment.go:
```diff
@@ -0,0 +1,405 @@
+			"flow_id": schema.StringAttribute{
+				CustomType:  customtypes.UUIDType{},
+				Description: "Flow ID (UUID) to associate deployment to",
+				Optional:    true,
```
aside from the user-provided inputs (name, id, account_id, workspace_id), should any attribute be `Optional` and not just read-only `Computed: true`?

mitchnielsen replied:
good call - I understand this better now, updated in 1b35a3d (but let me know if that doesn't match what you were thinking)

parkedwards commented on internal/provider/datasources/deployment.go:
```diff
@@ -0,0 +1,405 @@
+	}
+
+	// A deployment can be imported + read by either ID or Handle
+	// If both are set, we prefer the ID
```
👍

parkedwards reviewed (approved):
reusing the resource components is smart. just had a question about Optional values in the schema - im not 100% sure we need them, but down to be wrong

mitchnielsen committed:
Only use optional for configurable attributes

From the docs on Optional:

Optional indicates whether the practitioner can choose to enter a value
for this attribute or not. Optional and Required cannot both be true.

Only `name` and `id` (and `{workspace,account}_id`) are configurable, so
ensure those are marked Optional and remove that field from ther est of
them.

This also matches the example from the docs on implementing datasources:
https://developer.hashicorp.com/terraform/tutorials/providers-plugin-framework/providers-plugin-framework-data-source-read

github-actions[bot] committed:
Generate Terraform Docs
